### PR TITLE
Added Dutch National Flag algorithm #4636

### PR DESCRIPTION
**Pull request by KiranHipparagi**
### **Describe your change:**
Added Dutch National Flag algorithm

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by KiranHipparagi** [comment]
@cclauss @dhruvmanila please review

**Comment by cclauss** [comment]
Please format your code with psf/black as discussed in CONTRIBUTING.md.  `dnf_sort()` needs doctest including input with a -1, 3, 1.1, and "A" to see how the function deals with bad input.

**Comment by KiranHipparagi** [comment]
@cclauss The elements inside the sequence must contains only [0, 1, 2] values as per [dutch national flag](https://en.wikipedia.org/wiki/Dutch_national_flag_problem) standard.But in future we can add other features also..

**Comment by cclauss** [comment]
> The elements inside the sequence must contains only [0, 1, 2]

Yes, but the tests need to determine how the algorithm deals with bad data as discussed in CONTRIBUTING.md.

File: sorts/dnf_sort.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+We have to keep in mind that this algorithm can be implemented only on an array that has three unique elements.
+
+1)Time complexity is O(n).
+2)Space complexity is O(1).
+
+More info on: https://en.wikipedia.org/wiki/Dutch_national_flag_problem
+
+For doctests run following command:
+python -m doctest -v dnf_sort.py
+or
```
Comment (cclauss) [inline sorts/dnf_sort.py:19]:
```suggestion
```

File: sorts/dnf_sort.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+
+# Python program to sort an array with
+# 0, 1 and 2 in a single pass
+red = 0 # In this case this is the first color of the flag.
+white = 1 # This is the second color of the flag.
+blue = 2 # This is the third color of the flag.
+colors = [red, white, blue]
+# Function to sort array
+def dnf_sort( sequence:list, arr_size:int)->list:
```
Comment (cclauss) [inline sorts/dnf_sort.py:35]:
```suggestion
red = 0  # In this case this is the first color of the flag.
white = 1  # This is the second color of the flag.
blue = 2  # This is the third color of the flag.
colors = (red, white, blue)
# Function to sort array
def dutch_national_flag_sort(sequence: list, seq_size: int) -> list:
```

File: sorts/dnf_sort.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    """
+     Pure implementation of dnf sort algorithm in Python
```
Comment (cclauss) [inline sorts/dnf_sort.py:37]:
```suggestion
    Pure Python implementation of Dutch National Flag sort algorithm
```

File: sorts/dnf_sort.py (hunk 2 above)
Comment (cclauss) [reply sorts/dnf_sort.py:35]: Why do we ask for the size of `sequence` if we can obtain is by doing `len(sequence)`?  What happens if the caller says it is 3 but it is 5?  What happens if the caller says it is 7 but it is 5?

File: sorts/dnf_sort.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    :param data: 3 unique values (eg. 0,1,2) in an Array
```
Comment (cclauss) [inline sorts/dnf_sort.py:38]:
```suggestion
    :param data: 3 unique integer values (e.g., 0, 1, 2) in an sequence
```

File: sorts/dnf_sort.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    :return: the same collection in ascending order
+    Examples:
+    Input: dnf_sort([2, 1, 0, 0,1, 2],6)
+    Output: [0, 0, 1, 1, 2, 2]
+    Input: dnf_sort([0, 1, 1, 0, 1, 2, 1, 2, 0, 0, 0, 1],12)
+    Output: [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 2, 2]
+    """
+    if sequence is None: return
+    if len(sequence) <= 1: return
```
Comment (cclauss) [inline sorts/dnf_sort.py:47]: We promised in the function signature above that this function would return a list.
```suggestion
    if sequence is None:
        return []
    if len(sequence) <= 1:
        return []
```

File: sorts/dnf_sort.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            continue
+        if sequence[mid] == colors[1]:
+            mid = mid + 1
+            continue
+        if sequence[mid] == colors[2]:
+            sequence[mid], sequence[high] = sequence[high], sequence[mid]
+            high = high - 1
+            continue
+            
+        raise ValueError(f"The elements inside the sequence must contains only {colors} values")
```
Comment (cclauss) [inline sorts/dnf_sort.py:65]:
```suggestion
        if sequence[mid] == colors[0]:
            sequence[low], sequence[mid] = sequence[mid], sequence[low]
            low += 1
            mid += 1
        elif sequence[mid] == colors[1]:
            mid += 1
        elif sequence[mid] == colors[2]:
            sequence[mid], sequence[high] = sequence[high], sequence[mid]
            high -= 1
        else:
            raise ValueError(f"The elements inside the sequence must contains only {colors} values")    
```

**Comment by KiranHipparagi** [comment]
@cclauss soory this is my first contribution.i have gone through CONTRIBUTING.md and followed all the steps.Not getting what exactly needs to change. could you please give me link if you have anything so that i'll check and correct it. Thanks

File: sorts/dnf_sort.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        raise ValueError(f"The elements inside the sequence must contains only {colors} values")    
+    return sequence
+
+if __name__ == "__main__":
+    import doctest
+
+    doctest.testmod()
+
+    assert dnf_sort([1, 0, 0, 2, 1],5) == [0,0,1, 1, 2]
+    assert dnf_sort([0, 1, 1, 0, 1, 2, 1, 2, 0, 0, 0, 1],12) == [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 2, 2]
```
Comment (cclauss) [inline sorts/dnf_sort.py:74]:
```suggestion
```
These can be done as doctests which are both human and machine readable/testaable.

File: sorts/dnf_sort.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    user_input = input("Enter numbers separated by a comma:\n").strip()
+    unsorted = [int(item) for item in user_input.split(",")]
+print(f"{dnf_sort(unsorted,len(unsorted))}")
```
Comment (cclauss) [inline sorts/dnf_sort.py:77]:
```suggestion
    print(f"{dutch_national_flag_sort(unsorted, len(unsorted))}")
```

**Comment by cclauss** [comment]
$ ` mypy .`
```
sorts/dnf_sort.py:47: error: Return value expected
Found 1 error in 1 file (checked 318 source files)
```

**Comment by cclauss** [comment]
Click the `details` link next to the ❌es below to see what tests are failing.

**Review by cclauss — CHANGES_REQUESTED** [review]

File: sorts/dnf_sort.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+def dutch_national_flag_sort( sequence: list) -> list:
+    """
+     Pure Python implementation of Dutch National Flag sort algorithm
+    :param data: 3 unique integer values (e.g., 0, 1, 2) in an sequence
+    :return: the same collection in ascending order
+    Examples:
+    Input: dutch_national_flag_sort([2, 1, 0, 0,1, 2])
+    Output: [0, 0, 1, 1, 2, 2]
+    Input: dutch_national_flag_sort([0, 1, 1, 0, 1, 2, 1, 2, 0, 0, 0, 1])
+    Output: [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 2, 2]
```
Comment (cclauss) [inline sorts/dnf_sort.py:43]: Doctests https://docs.python.org/3/library/doctest.html
```suggestion

    >>> dutch_national_flag_sort([2, 1, 0, 0,1, 2])
    [0, 0, 1, 1, 2, 2]
    >>> dutch_national_flag_sort([0, 1, 1, 0, 1, 2, 1, 2, 0, 0, 0, 1])
    [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 2, 2]
```

**Comment by KiranHipparagi** [comment]
@cclauss thank you i'll make correction.

**Comment by KiranHipparagi** [comment]
@cclauss updated doctest.Thanks for the clarification.

File: sorts/dnf_sort.py (hunk 2 above)
Comment (KiranHipparagi) [reply sorts/dnf_sort.py]: @cclauss  yes  i have updated seq_size.please review and confirm

**Comment by cclauss** [comment]
https://github.com/TheAlgorithms/Python/pull/4639/checks?check_run_id=3473281739 is a :X below.  The `-` lines are your code and the `+` lines are how https://pypi.org/project/black would reformat your code.

All lines in your code should be 88 characters max or you will continue to see ❌es.

**Comment by KiranHipparagi** [comment]
> https://github.com/TheAlgorithms/Python/pull/4639/checks?check_run_id=3473281739 is a x below. The `-` lines are your code and the `+` lines are how https://pypi.org/project/black would reformat your code.
> 
> All lines in your code should be 88 characters max or you will continue to see xes.
#### updated file as per black code formatter.

**Comment by KiranHipparagi** [comment]
@cclauss please review.

**Review by cclauss — APPROVED** [review]
Awesome work!  Thanks for your persistence.  It is a cool algorithm.

**Pull request merged**